### ISSUE TITLE
Diagnostics Finished With Error

### ISSUE THREAD

**Issue by axoide** [codenvy/codenvy#1917]
After installation of Codenvy, the execution of the diagnostic tool ends with error during the "REST Call on Workspace Agent" test:

> Unable to perform call on http://codenvy.axoide.fr/32808_codenvy.axoide.fr/api : Status 401
> HTTP Status 401 - Authentication on machine failed, token DraCySiTn9aGyb... is invalid
> Hint: Workspace Agent is running but unable to connect. Please check HOST defined in the env file and the firewall settings.

The CODENVY_HOST parameter is correct in the .env file and the Ubuntu firewall default values have not been changed.

Which parameter should be changed to correct this error?

**Reproduction Steps:**

1. Install Codenvy 5.4.1 on Ubuntu 16.04 LTS
2. Start codenvy/cli
3. Click on "Diagnostics" button from the lower corner of the dashboard

**Codenvy version:** codenvy/cli:5.4.1

**OS and version:** Ubuntu 16.04 LTS

**Docker version:** 17.03.0-ce   

**Codenvy cli.log output:**

```
HTTP/1.0 200 OK
Content-type: application/octet-stream
Date: Tue, 14 Mar 2017 14:17:00 GMT
Connection: close
Accept-Ranges: bytes
Last-Modified: Tue, 08 Nov 2016 19:28:30 GMT
Content-length: 6

         conn (browser => ws):    [OK]
         conn (server => ws):     [OK]
fakeagent

INFO: (codenvy start):  Starting containers...
docker_compose --file="/data/instance/docker-compose-container.yml" -p="codenvy" up -d >> "/data/cli.log" 2>&1
Creating codenvy_postgres_1
Creating codenvy_nginx_1
Creating codenvy_rsyslog_1
Creating codenvy_agents_1
Creating codenvy_socat_1
Creating codenvy_registry_1
Creating codenvy_zookeeper_1
Creating codenvy_haproxy_1
Creating codenvy_swarm_1
Creating codenvy_codenvy_1
docker inspect -f '{{.Id}}' codenvy_codenvy_1
INFO: (codenvy start):  Services booting...
INFO: (codenvy start):  Server logs at "docker logs -f codenvy_codenvy_1"
DEBUG: Container with id 61c58a036cb0 is matching postgres service
DEBUG: Running state of container 61c58a036cb0 is true
DEBUG: Container with id 89c1423f3f3b is matching nginx service
DEBUG: Running state of container 89c1423f3f3b is true
DEBUG: Container with id f7bfaaf7b713 is matching agents service
DEBUG: Running state of container f7bfaaf7b713 is true
DEBUG: Container with id ed252d56277f is matching rsyslog service
DEBUG: Running state of container ed252d56277f is true
DEBUG: Container with id 9a943ba6c025 is matching registry service
DEBUG: Running state of container 9a943ba6c025 is true
DEBUG: Container with id bbcd2de6106d is matching zookeeper service
DEBUG: Running state of container bbcd2de6106d is true
DEBUG: Container with id 5b0f5c1c341d is matching haproxy service
DEBUG: Running state of container 5b0f5c1c341d is true
DEBUG: Container with id 51e6d7fc8b80 is matching socat service
DEBUG: Running state of container 51e6d7fc8b80 is true
DEBUG: Container with id 0ffeed0601da is matching swarm service
DEBUG: Running state of container 0ffeed0601da is true
DEBUG: Container with id 24f9d8ec9a19 is matching codenvy service
DEBUG: Running state of container 24f9d8ec9a19 is true
/usr/bin/curl  "-I -k codenvy.axoide.fr/api/ -s -o /dev/null --write-out %{http_code}"
200
DEBUG: Container with id 61c58a036cb0 is matching postgres service
DEBUG: Running state of container 61c58a036cb0 is true
DEBUG: Container with id 89c1423f3f3b is matching nginx service
DEBUG: Running state of container 89c1423f3f3b is true
DEBUG: Container with id f7bfaaf7b713 is matching agents service
DEBUG: Running state of container f7bfaaf7b713 is true
DEBUG: Container with id ed252d56277f is matching rsyslog service
DEBUG: Running state of container ed252d56277f is true
DEBUG: Container with id 9a943ba6c025 is matching registry service
DEBUG: Running state of container 9a943ba6c025 is true
DEBUG: Container with id bbcd2de6106d is matching zookeeper service
DEBUG: Running state of container bbcd2de6106d is true
DEBUG: Container with id 5b0f5c1c341d is matching haproxy service
DEBUG: Running state of container 5b0f5c1c341d is true
DEBUG: Container with id 51e6d7fc8b80 is matching socat service
DEBUG: Running state of container 51e6d7fc8b80 is true
DEBUG: Container with id 0ffeed0601da is matching swarm service
DEBUG: Running state of container 0ffeed0601da is true
DEBUG: Container with id 24f9d8ec9a19 is matching codenvy service
DEBUG: Running state of container 24f9d8ec9a19 is true
docker inspect -f '{{.Id}}' codenvy_codenvy_1
/usr/bin/curl  "-I -k codenvy.axoide.fr/api/ -s -o /dev/null --write-out %{http_code}"
200
INFO: (codenvy start):  Postflight checks
/usr/bin/curl  "-s codenvy.axoide.fr:23750/info -o /dev/null --write-out %{http_code}"
         (codenvy.axoide.fr:23750/info):  [OK]

docker inspect -f '{{.Id}}' codenvy_codenvy_1
/usr/bin/curl  "-I -k codenvy.axoide.fr/api/ -s -o /dev/null --write-out %{http_code}"
200
INFO: (codenvy start):  Booted and reachable
INFO: (codenvy start):  Ver: 5.4.1
INFO: (codenvy start):  Use: http://codenvy.axoide.fr:80
INFO: (codenvy start):  API: http://codenvy.axoide.fr:80/swagger
Tue Mar 14 17:12:50 UTC 2017
/usr/bin/curl  "-I -k https://hub.docker.com -s -o /dev/null --write-out %{http_code}"
Checking registry for version '5.4.1' images
eval BOOTSTRAP_IMAGE_ALPINE=alpine:3.4
eval BOOTSTRAP_IMAGE_CHEIP=eclipse/che-ip:5.4.1
eval IMAGE_HAPROXY=haproxy:1.5.18-alpine
eval IMAGE_NGINX=nginx:1.10-alpine
eval IMAGE_SWARM=swarm:1.2.6
eval IMAGE_REGISTRY=registry:2.5.0
eval IMAGE_POSTGRES=postgres:9.5.4
eval IMAGE_COMPOSE=docker/compose:1.8.1
eval IMAGE_ZOOKEEPER=zookeeper:3.4.9
eval IMAGE_SOCAT=codenvy/socat:5.4.1
eval IMAGE_INIT=codenvy/init:5.4.1
eval IMAGE_AGENTS=codenvy/agents:5.4.1
eval IMAGE_CODENVY=codenvy/codenvy:5.4.1
eval IMAGE_RSYSLOG=codenvy/rsyslog:5.4.1
eval UTILITY_IMAGE_CHEACTION=eclipse/che-action:5.4.1
eval UTILITY_IMAGE_CHEDIR=eclipse/che-dir:5.4.1
eval UTILITY_IMAGE_CHETEST=eclipse/che-test:5.4.1
eval UTILITY_IMAGE_CHEMOUNT=eclipse/che-mount:5.4.1
INFO: (codenvy cli):  5.4.1 - using docker 17.03.0-ce / native
INFO: (codenvy cli):  'CODENVY_HOST=codenvy.axoide.fr' is != discovered IP '163.172.59.212'
/usr/bin/curl  "-s https://hub.docker.com/v2/repositories/codenvy/cli/tags/"
DEBUG: cmd_info

CLI:
 TTY:            true
 Daemon:         /var/run/docker.sock
 Image:          codenvy/cli
 Version:        5.4.1
 Command:        info
 Parameters:     info
Mounts:
 /data:          /home/borgognoc/codenvy
 /data/instance: not set
 /data/backup:   not set
 /repo:          not set
 /assembly:      not set
 /sync:          not set
 /unison:        not set
 /chedir:        not set
System:
 Docker:         native
 Proxy:          not set
Internal:
 CODENVY_VERSION:    5.4.1
 CODENVY_HOST:       codenvy.axoide.fr
 CODENVY_INSTANCE:   /home/borgognoc/codenvy/instance
 CODENVY_CONFIG:     /home/borgognoc/codenvy
 CODENVY_BACKUP:     /home/borgognoc/codenvy/backup
 CODENVY_REGISTRY:   /version
 CODENVY_DEBUG:      false
 IP Detection:   163.172.59.212
 Initialized:    true
Image Registry:
 IMAGE_HAPROXY=haproxy:1.5.18-alpine
 IMAGE_NGINX=nginx:1.10-alpine
 IMAGE_SWARM=swarm:1.2.6
 IMAGE_REGISTRY=registry:2.5.0
 IMAGE_POSTGRES=postgres:9.5.4
 IMAGE_COMPOSE=docker/compose:1.8.1
 IMAGE_ZOOKEEPER=zookeeper:3.4.9
 IMAGE_SOCAT=codenvy/socat:5.4.1
 IMAGE_INIT=codenvy/init:5.4.1
 IMAGE_AGENTS=codenvy/agents:5.4.1
 IMAGE_CODENVY=codenvy/codenvy:5.4.1
 IMAGE_RSYSLOG=codenvy/rsyslog:5.4.1
 BOOTSTRAP_IMAGE_ALPINE=alpine:3.4
 BOOTSTRAP_IMAGE_CHEIP=eclipse/che-ip:5.4.1
 UTILITY_IMAGE_CHEACTION=eclipse/che-action:5.4.1
 UTILITY_IMAGE_CHEDIR=eclipse/che-dir:5.4.1
 UTILITY_IMAGE_CHETEST=eclipse/che-test:5.4.1
 UTILITY_IMAGE_CHEMOUNT=eclipse/che-mount:5.4.1
codenvy.env:
 CODENVY_HOST=codenvy.axoide.fr
 CODENVY_ADMIN_EMAIL=claude.borgogno@axoide.com
 CODENVY_SWARM_NODES=codenvy.axoide.fr:23750
 DOCKER_HOST=tcp://codenvy.axoide.fr:2375
 CHE_DOCKER_IP_EXTERNAL=163.172.59.212
```

**Comment by TylerJewell** [comment]
This is likely a firewall issue.  Are the ephemeral ports 32768+ open between the browser and where Codenvy is installed?

**Comment by axoide** [comment]
Ephemeral ports 32768+ are open.
The ephemeral ports range specified in /proc/sys/net/ipv4/ip_local_port_range is:
`32768   60999`

**Comment by TylerJewell** [comment]
@benoitf - ideas about other root causes of this test failing?

**Comment by axoide** [comment]
Here is a screenshot:

![screen](https://cloud.githubusercontent.com/assets/2506915/23921960/de40350a-0900-11e7-9e78-7d454d787cba.png)

**Comment by benoitf** [comment]
I need to check on a ubuntu install. It seems the token is invalid so maybe there was a change on API

@axoide are you able to start and use a workspace ?

**Comment by axoide** [comment]
I can start and succesfully use a workspace with a blank projet, but creating a workspace with a java-mysql (petclinic) project fails and ends with the following error:
`Could not start workspace wksp-ket3. Reason: Start of environment 'wksp-ket3' failed. Error: Fail launching agent Terminal. Workspace ID: workspace7fdpwz3ugdcutru4`

**Comment by TylerJewell** [comment]
Are you behind any kind of firewall - the terminal is in a different port range that is below 32,768.

**Comment by axoide** [comment]
`iptables -L` displays:

```
Chain INPUT (policy ACCEPT)
target     prot opt source               destination         

Chain FORWARD (policy DROP)
target     prot opt source               destination         
DOCKER-ISOLATION  all  --  anywhere             anywhere            
DOCKER     all  --  anywhere             anywhere            
ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
ACCEPT     all  --  anywhere             anywhere            
ACCEPT     all  --  anywhere             anywhere            
DOCKER     all  --  anywhere             anywhere            
ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
ACCEPT     all  --  anywhere             anywhere            
ACCEPT     all  --  anywhere             anywhere            
DOCKER     all  --  anywhere             anywhere            
ACCEPT     all  --  anywhere             anywhere             ctstate RELATED,ESTABLISHED
ACCEPT     all  --  anywhere             anywhere            
ACCEPT     all  --  anywhere             anywhere            

Chain OUTPUT (policy ACCEPT)
target     prot opt source               destination         

Chain DOCKER (3 references)
target     prot opt source               destination         
ACCEPT     tcp  --  anywhere             172.18.0.6           tcp dpt:shell
ACCEPT     tcp  --  anywhere             172.18.0.4           tcp dpt:2375
ACCEPT     tcp  --  anywhere             172.18.0.7           tcp dpt:5000
ACCEPT     tcp  --  anywhere             172.18.0.8           tcp dpt:2181
ACCEPT     tcp  --  anywhere             172.18.0.10          tcp dpt:2375
ACCEPT     tcp  --  anywhere             172.18.0.11          tcp dpt:32101
ACCEPT     tcp  --  anywhere             172.18.0.11          tcp dpt:32001
ACCEPT     tcp  --  anywhere             172.18.0.9           tcp dpt:https
ACCEPT     tcp  --  anywhere             172.18.0.9           tcp dpt:http
ACCEPT     tcp  --  anywhere             172.20.0.2           tcp dpt:4411
ACCEPT     tcp  --  anywhere             172.20.0.2           tcp dpt:4401
ACCEPT     tcp  --  anywhere             172.20.0.2           tcp dpt:ssh

Chain DOCKER-ISOLATION (1 references)
target     prot opt source               destination         
DROP       all  --  anywhere             anywhere            
DROP       all  --  anywhere             anywhere            
DROP       all  --  anywhere             anywhere            
DROP       all  --  anywhere             anywhere            
DROP       all  --  anywhere             anywhere            
DROP       all  --  anywhere             anywhere            
RETURN     all  --  anywhere             anywhere
```

**Comment by TylerJewell** [comment]
That looks clean.  Let's see what some of the other engineers think.

**Comment by eivantsov** [comment]
@TylerJewell If a normal workspace can be started and java-mysql can't, it's 100% an issue with Terminal agent downloading dependencies in mysql container that is Debian based. Debian repos are acting up sometimes. Also, it depends on what mirrors are used (thus your local matters). There's a 1 min timeout for a terminal agent and if it fails to install itself in 1 minute a workspace does not start (workspace environment to be exact).

This is what happens when a terminal agent is being launched:

https://github.com/eclipse/che/blob/master/agents/go-agents/src/main/resources/org.eclipse.che.terminal.script.sh#L59-L65
]
Official mysql image that we use does not have curl. Before installing curl, `sudo apt-get update` is run:

```
root@ubuntu-2gb-fra1-01:~/che-dockerfiles/recipes# docker run -ti mysql sh
# curl
sh: 1: curl: not found
```

To fix it, we either use our own image with all the stuff that agents may require, or use a recipe in a composefile - `FROM mysql\nRUN apt-get update && apt-get install curl`

cURL is used to download terminal from `agents` container with https server running in there.

**Comment by TylerJewell** [comment]
@eivantsov - that is a very thorough analysis.  So, does the diagnostic utility make use of the java / mysql image for its tests?  You jumped right as that being the culprit, so that would make sense if it is used in this way.

@eivantsov - on the fix for the updated compose file - is this something that we would do over a PR, or is it another way to handle that?

@axoide - are you able to create just a plain Java project as well  (you confirmed a blank one earlier)?

**Comment by benoitf** [comment]
@tylerjewell we're not we use openjdk image

**Comment by TylerJewell** [comment]
@benoitf - is the container that you are using to test the workspace a debian-based container instead of ubuntu or alpine?  I think you went with something that was micro size for speed.

**Comment by benoitf** [comment]
It is using alpine

**Comment by TylerJewell** [comment]
Ok, then can someone help me connect Eugene's statements saying that the issue is Debian-based vs. your statements?  Perhaps my description of the problem and / or bug is not accurate?

**Comment by benoitf** [comment]
I think it's a bug in diagnostic tool

**Comment by TylerJewell** [comment]
Oh ... now the truth comes out.  I think you were hiding from me on that one :)

**Comment by eivantsov** [comment]
Yes, diagnostic check failure and java-mysql failure are two different failures.

**Comment by benoitf** [comment]
Fixed in current master and will be available in 5.6.0
